### PR TITLE
New 'set background effect' block that offers extended palettes/backgrounds (AI Dance Party)

### PR DESCRIPTION
**Pull request by fisher-alice**
This PR pairs with https://github.com/code-dot-org/dance-party/pull/658 which adds the Poetry lab color palettes to Dance Lab.

This PR adds a new block `Dancelab_setBackgroundEffectWithPaletteAI` in Dance Lab that will only be used in the AI Dance Party progression. This is a duplicate of the `Dancelab_setBackgroundEffectWithPalette` block but will offer 12 new color palettes and 4 new background effects.

This PR also adds these options to the call to `doAI` and if the user chooses to 'Convert' the AI block, the `Dancelab_setBackgroundEffectWithPaletteAI` block will be generated along with the `Dancelab_setForegroundEffect` block.

Here is a screencast video of the AI block converted to the `Dancelab_setBackgroundEffectWithPaletteAI` block with the extended dropdown menu. It also shows at the end as a comparison the `Dancelab_setBackgroundEffectWithPalette` block that contains only the original dropdown options.

https://github.com/code-dot-org/code-dot-org/assets/107423305/e9355efa-1b8f-4e31-b8a3-ee267f0b03ad

## Links
[Trello](https://trello.com/c/4qnud1Y9/51-make-poetry-background-palettes-available)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/dance/ai/utils.ts [hunk 1]
```diff
@@ -2,9 +2,23 @@ import HttpClient from '@cdo/apps/util/HttpClient';
...
+  const oldDanceBackgroundEffects =
+    'circles, color_cycle, diamonds, disco_ball, fireworks, swirl, kaleidoscope, lasers, splatter, rainbow, snowflakes, galaxy, sparkles, spiral, disco, stars';
+  const newDanceBackgroundEffects =
```
Comment (fisher-alice) [inline apps/src/dance/ai/utils.ts]: We have a [ticket](https://trello.com/c/QpakSqaK/55-add-helper-function-to-return-lists-of-background-foreground-effects-and-palettes) to create a static file from ‘dashboard/config/blocks/Dancelab/Dancelab_setBackgroundEffectWithPaletteAI'. For now, we set the lists here.

File: apps/src/dance/constants.js [hunk 2]
```diff
@@ -39,3 +39,18 @@ export const DancelabReservedWords = [
 
 export const ASSET_BASE =
   'https://curriculum.code.org/images/sprites/dance_20191106/';
+
+export const oldDanceBackgroundEffects =
+  'circles, color_cycle, diamonds, disco_ball, fireworks, swirl, kaleidoscope, lasers, splatter, rainbow, snowflakes, galaxy, sparkles, spiral, disco, stars';
+
+export const newDanceBackgroundEffects =
```
Comment (jmkulwik) [inline apps/src/dance/constants.js]: NIT: new and old feels a little odd here. For example, if we add more next year, do we name them "newerDanceBackgroundEffects"

Notes from our brainstorm: Perhaps "commonDanceBackgroundEffects" and "aiModuleDanceBackgroundEffects" or perhaps "v2ModuleDanceBackgroundEffects"

Comment (jmkulwik) [reply apps/src/dance/constants.js]: Maybe even "danceBackgroundEffectsExtras"

Comment (fisher-alice) [reply apps/src/dance/constants.js]: Thanks! Renamed constants - lmk what you think.

File: apps/src/dance/Dance.js [hunk 3]
```diff
@@ -628,12 +628,9 @@ Dance.prototype.execute = async function () {
...
 
   const songMetadata = await this.songMetadataPromise;
-  const userBlockTypes = [];
-  Blockly.getMainWorkspace()
+  const userBlockTypes = Blockly.getMainWorkspace()
     .getAllBlocks()
-    .forEach(block => {
-      userBlockTypes.push(block.type);
-    });
+    .map(block => block.type);
```
Comment (fisher-alice) [inline apps/src/dance/Dance.js:633]: Unrelated cleanup. @breville

File: apps/src/dance/constants.js [hunk 4]
```diff
@@ -39,3 +39,18 @@ export const DancelabReservedWords = [
 
 export const ASSET_BASE =
   'https://curriculum.code.org/images/sprites/dance_20191106/';
+
+export const commonDanceBackgroundEffects =
```
Comment (bencodeorg) [inline apps/src/dance/constants.js]: A few nits:
- are these expected to be used multiple places? Might be easier to keep where they're used if only used once?
- all caps for variable names?
- might be more readable to put them in an array (with one per line), then use them via array.join()

Regardless of of whether you do any of these, this is a nice bit of cleanup 👍

**Review by bencodeorg — APPROVED** [review]
LGTM! I'm not super familiar with the block adding process, this won't show up in any existing dance party levels / project levels / etc, right? I think per your PR description its only for new AI levels, but just confirming

**Comment by fisher-alice** [comment]
> LGTM! I'm not super familiar with the block adding process, this won't show up in any existing dance party levels / project levels / etc, right? I think per your PR description its only for new AI levels, but just confirming

Yes - that's correct. We only use the block when it's generated from the AI block.

File: apps/src/dance/constants.js (hunk 4 above)
Comment (fisher-alice) [reply apps/src/dance/constants.js]: I see that we have a duplicate list used in @tshaffercodeorg's PR https://github.com/code-dot-org/code-dot-org/pull/54299 in `HoC2023AiGenerateWeights.py` as @sanchitmalhotra126 pointed out.
I think I'll keep them in defined in this file for now to de-clutter `utils.ts`. Definitely should have one source of truth for these - task is created for this.

Also - I converted to arrays and used all caps. Thanks!
